source/event_stream_rpc_server.c

```diff
@@ -0,0 +1,906 @@
+}
+
+static void s_destroy_server(struct aws_event_stream_rpc_server_listener *server) {
+    if (server) {
+        /* the memory for this is cleaned up in the listener shutdown complete callback. */
+        aws_server_bootstrap_destroy_socket_listener(server->bootstrap, server->listener);
+    }
+}
+
+void aws_event_stream_rpc_server_listener_release(struct aws_event_stream_rpc_server_listener *server) {
+    size_t ref_count = aws_atomic_fetch_sub_explicit(&server->ref_count, 1, aws_memory_order_seq_cst);
+
+    if (ref_count == 1) {
+        s_destroy_server(server);
+    }
+}
+
+struct event_stream_connection_send_message_args {
+    struct aws_allocator *allocator;
+    struct aws_event_stream_message message;
+    enum aws_event_stream_rpc_message_type message_type;
+    struct aws_event_stream_rpc_connection *connection;
+    struct aws_event_stream_rpc_server_continuation_token *continuation;
+    aws_event_stream_rpc_server_message_flush_fn *flush_fn;
+    void *user_data;
+    bool end_stream;
+    bool terminate_connection;
+};
+
+static void s_on_protocol_message_written_fn(
+    struct aws_event_stream_message *message,
+    int error_code,
+    void *user_data) {
+    (void)message;
+
+    struct event_stream_connection_send_message_args *message_args = user_data;
+
+    if (message_args->message_type == AWS_EVENT_STREAM_RPC_MESSAGE_TYPE_CONNECT_ACK) {
+        aws_atomic_store_int(&message_args->connection->handshake_complete, 2);
+    }
+
+    if (message_args->end_stream) {
+        AWS_FATAL_ASSERT(message_args->continuation && "end stream flag was set but it wasn't on a continuation");
+        aws_atomic_store_int(&message_args->continuation->is_closed, 1U);
+        aws_hash_table_remove(
+            &message_args->connection->continuation_table, &message_args->continuation->stream_id, NULL, NULL);
+    }
+
+    message_args->flush_fn(error_code, message_args->user_data);
+
+    if (message_args->terminate_connection) {
+        aws_event_stream_rpc_server_connection_close(message_args->connection, AWS_ERROR_SUCCESS);
+    }
+
+    aws_event_stream_rpc_server_connection_release(message_args->connection);
+
+    if (message_args->continuation) {
+        aws_event_stream_rpc_server_continuation_release(message_args->continuation);
+    }
+
+    aws_event_stream_message_clean_up(&message_args->message);
+    aws_mem_release(message_args->allocator, message_args);
+}
+
+static int s_send_protocol_message(
+    struct aws_event_stream_rpc_connection *connection,
+    struct aws_event_stream_rpc_server_continuation_token *continuation,
+    const struct aws_event_stream_rpc_message_args *message_args,
+    int32_t stream_id,
+    aws_event_stream_rpc_server_message_flush_fn *flush_fn,
+    void *user_data) {
+
+    size_t connect_handshake_completion = aws_atomic_load_int(&connection->handshake_complete);
+
+    if (connect_handshake_completion != 2 &&
```

Comment:
Why 2? I'm sure I'll figure it out if I read more, but a comment would help too.

Reply:
I added them elsewhere, will add here as well.